coderd/database/pubsub/latency.go

```diff
@@ -0,0 +1,55 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+)
+
+var channelID uuid.UUID
+
+// Create a new pubsub channel UUID per coderd instance so that multiple replicas do not clash when performing latency
+// measurements, and only create one UUID per instance (and not request) to limit the number of notification channels
+// that need to be maintained by the Pubsub implementation.
+func init() {
+	channelID = uuid.New()
+}
+
+// MeasureLatency takes a given Pubsub implementation, publishes a message & immediately receives it, and returns the
+// observed latency.
+func MeasureLatency(ctx context.Context, p Pubsub) (send float64, recv float64, err error) {
+	var (
+		start time.Time
+		res   = make(chan float64, 1)
+	)
+
+	cancel, err := p.Subscribe(latencyChannelName(), func(ctx context.Context, _ []byte) {
+		res <- time.Since(start).Seconds()
```

Comment:
having each measurement be an identical message (and ignoring it) could lead to previously timed-out measurements getting processed here.  That could lead to very strange results such as a ping arriving before `start` is set, which will result in the `recv` latency being two millennia 🙀

Reply:
Good call! OK, I'll refactor to generate a unique message and only mark the latency roundtrip done when I observe that.

Reply:
Let me know what you think of 65f57b1cc2842f40fcd5ed4f775e60ce18fb12b7 please

Reply:
49d2002e7c2091dba6a98b6e60c871681e581913 supercedes the above